### ISSUE TITLE
SO(3) / SE(3) with matrix storage for optimization in Ceres

### ISSUE THREAD

**Issue by NikolausDemmel** [strasdat/Sophus#185]
**TL/DR:**
1) Other than possible increased performance (avoiding computation of q.matrix() and the corresponding derivative), there should be very little difference between a rotation-matrix-backed or quaternion-back SE3 parameterizing when doing optimization with Ceres, right?
2) If I wanted to create a version of Sophus' SO(3) and SE(3) with rotation matrix instead of quaternion, there is currently no nice way to reuse the API, but instead I'd have to effectively create a copy of everything, right?

**Full version:**

I want to implement an optimization with SE(3) state variables in Ceres, that mimicks in its behaviour exactly an existing implementation (that doesn't use ceres). The existing implementation stores rotation matrix and tranlsation vector and uses a local Lie Algebra parameterization with the matrix exponential as a retraction.

My initial thought was, that I need to create a SE(3) class backed by rotation matrix storage, since in the local parameterization for Ceres you need to specify an update and Jacobian that is specific to the quaternion storage (in the case of Sophus). But it would be much nicer to reuse Sophus' SE(3) class of course.

This is related to a previous discussion on "internal generators" with @strasdat (see https://github.com/strasdat/Sophus/issues/118#issuecomment-350326780):

>> While SE(3) / SO(3) etc. are pretty much covered by standard textbooks, I'm not aware of a good reference when it comes to the concept I call "internal generators". This concept arises by the fact that we represent SO(3) internally as SU(2).d

> Is there really something fundamental going on here? It is simply a matter of representing rotations in different and equivalent ways. And its fine to mix and match, as long as you are consistent. E.g. for manifold optimization, what Jacobians you need depends on whether you differentiate down to the quaternion representation, or stop at the direction cosine matrix, and at the same time whether you apply the update directly to the quaternion, or whether you update an intermediate direction cosine matrix and then convert the result to a quaternion.

However, on a second thought, this should not matter after all. Let's consider what would be the difference in the local parameterization for a rotation matrix implementation:

**Pose update:** The `Plus` method, like in [Sophus' implementation](https://github.com/strasdat/Sophus/blob/master/test/ceres/local_parameterization_se3.hpp), would still boil down to `T * SE3::exp(delta)`, so as long as Sophus' exponential is compatible with the true matrix exponential (in the sense that a given `delta` results in the same rigid body transformation, just represented differently). And this is in fact the case. (Note that this is not completely trivial. For example Ceres comes with a [local parameterization for SO(3)](https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/local_parameterization.cc#L163) using quaternions that is slightly different from the rotation matrix version, since it's exponential is missing a factor of 1/2 compared to Sophus.)

**ComputeJacobian:** Essentially this computes a Jacobian "Dx (T * exp(x))" at x=0, which in Sophus' case is 7x6 Jacobian and in a 3x3 rotation matrix case would be 12x6 Jacobian J_R for SE(3). Let us consider only the rotation part, i.e. the 4x3 Jacobian J_q for the quaternion representation and the 9x3 Jacobian J_R for the rotation matrix version. Let's further consider a residual function f(R), that depends on the rotation matrix. With the 3x3 matrix parameterization, Ceres would use autodiff on f and concatenate that with J_R as (D_R f(R)) * J_R. In the case of a quaternion parameterization, for the residual f(q.matrix()), ceres would compute the derivative as (D_q f(q.matrix())) * J_q. However, given the chain rule, both Jacobians should be exactly identical (except for numerical differences, since we compute them differently).

So my questions are:
1. Does it make sense to say that I can use Sophos' SE(3) implementation with internal SU(2) with together with ceres and get an optimization that runs exactly as if I would use the corresponding local parameterization for an implementation storing the rotation as a 3x3 matrix? In the sense that all Jacobians and pose-updates are analytically identical (assuming numerical differences of the conversions between the representation or when concatenating Jacobians don't play a role and can be assumed "exact")?
2. The reason I still might want to use a 3x3 matrix backed storage in the end is performance, since if I use the conversion `q.matrix()` a lot in my ceres residuals, it is slower to compute the residuals and also the Jacobians. It seems there is no way to easily make an alternative SO(3) implementation work with the existing SE(3) class (with some template magic)? Even the SO3Base class assumes a quaternion storage backing in the end? Therefore I would need to create a copy of all the SO(3) and SE(3) related classes and replace quaternions with 3x3 matrices?

**Comment by ojwoodford** [comment]
You can cache the matrix form inside the residual evaluator, and reset the cache each iteration of ceres using the evaluation callback. The issue with using the matrix form as a ceres parameter block is that you will have 9 variables instead of 4 for the rotation, which will increase the amount of computation you need to do (if using autodiff). The fact is, there should only need to be 3 rotation variables, so this is a deficiency in the way ceres implements local parameterizations. You can get round this by doing your own differentiation and filling the rest of your Jacobian with zeros, then implementing a special "Plus" method for the local parameterization. I have implemented all that of what I have described, and it works well. However, I cannot share the code.

**Comment by NikolausDemmel** [comment]
> You can cache the matrix form inside the residual evaluator, and reset the cache each iteration of ceres using the evaluation callback.

Good idea, thanks. However this will not make jacobians faster (at least when using autodiff).

> The issue with using the matrix form as a ceres parameter block is that you will have 9 variables instead of 4 for the rotation, which will increase the amount of computation you need to do (if using autodiff). The fact is, there should only need to be 3 rotation variables, so this is a deficiency in the way ceres implements local parameterizations. You can get round this by doing your own differentiation and filling the rest of your Jacobian with zeros, then implementing a special "Plus" method for the local parameterization. I have implemented all that of what I have described, and it works well. 

Yes, this is the same no matter which storage I use though. And it doesn't matter if my residual does actually depend on the full rotation matrix, and not for example on a rotated point, which is your case I assume. I have read about this way to work around the ceres limitation on the ceres mailing list. Thanks for pointing it out for reference.

> However, I cannot share the code.

No problem. One question: Have you actually profiled the performance gain from using this trick? That would be very interesting if you can share (even just rough numbers). I mean manually implemented Jacobains with the normal ceres local parameterization vs manually implemented Jacobians with this special local parameterizations with 0 entries that you mention. This would show the slowdown from the Ceres' API limitation.

**Comment by NikolausDemmel** [comment]
> Good idea, thanks. However this will not make jacobians faster (at least when using autodiff).

Oh I guess it would only work when not using autodiff anyway... forget what I said.

**Comment by ojwoodford** [comment]
> Good idea, thanks. However this will not make jacobians faster (at least when using autodiff).
> Oh I guess it would only work when not using autodiff anyway... forget what I said.

It does work when using autodiff. You cache the Jet variable. It should therefore make the Jacobians faster, whether or not you use autodiff, **if** you use the rotation/pose in many residual blocks. We did profile this approach (using autodiff, not the hand computed 3 parameter derivative, which should be faster still) and it gave a considerable speed boost. Note the boost you get is problem specific, so I won't bother to dig out the numbers (if I even have them saved).

**Comment by ojwoodford** [comment]
> One question: Have you actually profiled the performance gain from using this trick? That would be very interesting if you can share (even just rough numbers). I mean manually implemented Jacobains with the normal ceres local parameterization vs manually implemented Jacobians with this special local parameterizations with 0 entries that you mention. This would show the slowdown from the Ceres' API limitation.

No, I haven't. But it can be seen trivially that if your residual block has 10 variables, and you remove 1, you'll get an approximately 10% speed boost by reducing the number of Jacobian elements to compute by 10%. So the speed boost depends on the total number of variables in a residual block.

Actually, if you are using relative poses between two input pose variables you can also share the jacobian computation between them. That's another trick.

**Comment by NikolausDemmel** [comment]
> It does work when using autodiff. You cache the Jet variable. It should therefore make the Jacobians faster, whether or not you use autodiff, if you use the rotation/pose in many residual blocks. 

Ah cool. Yes that makes sense, thanks! I'll have to give it a try for how much difference that makes. That might mean there is even less reason to use a 3x3 rotation-matrix-backed parameterization, since as described in my OP, performance might be the only reason in certain cases, and the caching approach sounds like it **could** make the performance difference to a quaternion-backed paraemterization neglegible.
